python/ray/tune/integration/kubernetes.py

```diff
@@ -0,0 +1,143 @@
+        self.worker_node = None
+
+        sync_client = sync_client or KubernetesSyncClient(
+            namespace=self.__class__._namespace)
+
+        super(NodeSyncer, self).__init__(local_dir, remote_dir, sync_client)
+
+    def set_worker_ip(self, worker_ip):
+        self.worker_ip = worker_ip
+        self.worker_node = self._get_kubernetes_node_by_ip(worker_ip)
+
+    def _get_kubernetes_node_by_ip(self, node_ip):
+        """Return node name by internal or external IP"""
+        kubernetes.config.load_incluster_config()
+        api = kubernetes.client.CoreV1Api()
+        pods = api.list_namespaced_pod(self._namespace)
+        for pod in pods.items:
+            if pod.status.host_ip == node_ip or \
+               pod.status.pod_ip == node_ip:
+                return pod.metadata.name
+
+        logger.error(
+            "Could not find Kubernetes pod name for IP {}".format(node_ip))
+        return None
+
+    @property
+    def _remote_path(self):
+        return (self.worker_node, self._remote_dir)
+
+
+class KubernetesSyncClient(SyncClient):
+    """KubernetesSyncClient to be used by KubernetesSyncer.
+
+    This client takes care of executing the synchronization
+    commands for Kubernetes clients. In its ``sync_down`` and
+    ``sync_up`` commands, it expects tuples for the source
+    and target, respectively, for compatibility with the
+    KubernetesCommandRunner.
+
+    Args:
+        namespace (str): Namespace in which the pods live.
+        process_runner: How commands should be called.
+            Defaults to ``subprocess``.
+
+    """
+
+    def __init__(self, namespace, process_runner=subprocess):
+        self.namespace = namespace
+        self._process_runner = process_runner
+        self._command_runners = {}
+
+    def _create_command_runner(self, node_id):
+        """Create a command runner for one Kubernetes node"""
+        return KubernetesCommandRunner(
+            log_prefix="KubernetesSyncClient: {}:".format(node_id),
+            namespace=self.namespace,
+            node_id=node_id,
+            auth_config=None,
+            process_runner=self._process_runner)
+
+    def _get_command_runner(self, node_id):
+        """Create command runner if it doesn't exist"""
+        if node_id not in self._command_runners:
+            command_runner = self._create_command_runner(node_id)
+            self._command_runners[node_id] = command_runner
+        return self._command_runners[node_id]
+
+    def sync_up(self, source, target):
+        """Here target is a tuple (target_node, target_dir)"""
+        target_node, target_dir = target
+
+        command_runner = self._get_command_runner(target_node)
+        command_runner.run_rsync_up(source, target_dir)
+        return True
+
+    def sync_down(self, source, target):
+        """Here source is a tuple (source_node, source_dir)"""
+        source_node, source_dir = source
+
+        command_runner = self._get_command_runner(source_node)
```

Comment:
question - wouldn't you want to use the command runner that is already pointing at the target (rather than the one pointing at source), and then call "run_rsync_down"?

Reply:
The `sync_down` and then `run_rsync_down` method is called on the head node in `Trial.on_checkpoint()`. What you suggest would need to call the `run_rsync_up` command on the worker node (because the source dir is local), but currently the rest of the syncing management doesn't work that way.

Reply:
Hmm, wait I'm not sure if I understand what you're saying. Can you tell me what's wrong with my logic as presented below: 

You will always be invoking the syncer inside the the driver/head node process. 
You want to be able to do the following:
```
HEAD -rsync_up-> NODE  # uploads files to node
HEAD <-rsync_down- NODE  # downloads files from nodes.
```
This should be achievable by using the same CommandRunner and calling cmd_runner.run_rsync_up and run_rsync_down.

As an implementation detail, this manifests itself as:

```
# rsync-up
KUBECTL_RSYNC local_dir/ node@ip:remote_target/

# rsync-down 
KUBECTL_RSYNC node@ip:remote_dir/ local_target/
```

Reply:
Ah I think the confusion comes from the naming of source and target. The `NodeSyncer` calls the methods like so:

```
                result = self.sync_client.sync_up(self._local_dir,
                                                  self._remote_path)
```
and
```
                result = self.sync_client.sync_down(self._remote_path,
                                                    self._local_dir)
```

`remote_path` is the `worker_node, worker_path` tuple for the KubernetesCommand. So in both cases we actually do use the same command runner, but in one method it is the source and in one the target.

Your logic is completely right, and what you describe is what actually should happen. This is exactly what the test in this PR is trying to cover.